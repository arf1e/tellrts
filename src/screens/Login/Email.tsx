import React from 'react';
import {useTranslation} from 'react-i18next';
import {View} from 'react-native';
import Reanimated, {FadeInLeft, FadeOutLeft} from 'react-native-reanimated';
import PrimaryButton from '../../components/Buttons';
import ErrorDisplay from '../../components/ErrorDisplay';
import Field from '../../components/Field';
import {BodyCopy} from '../../components/Typography';
import LoginStyles from './Login.styles';

const AnimatedView = Reanimated.createAnimatedComponent(View);

interface Props {
  value: string;
  handleChange: (field: string) => any;
  handleSubmit: (e?: React.FormEvent<HTMLFormElement> | undefined) => void;
  error?: string;
  loading?: boolean;
  isInvalid?: boolean;
}

const Email = ({
  value,
  handleChange,
  handleSubmit,
  error,
  loading,
  isInvalid,
}: Props) => {
  const {t} = useTranslation();
  return (
    <AnimatedView
      entering={FadeInLeft.duration(320).delay(120)}
      exiting={FadeOutLeft.duration(320).damping(5000)}>
      <BodyCopy style={LoginStyles.fieldTitle}>
        {t('login.form.emailFieldTitle')}
      </BodyCopy>
      <Field
        defaultValue={value}
        autoCapitalize="none"
        keyboardType="email-address"
        placeholder={t('login.form.emailPlaceholder')}
        style={LoginStyles.emailField}
        onChangeText={handleChange}
        isInvalid={isInvalid}
      />
      {error && <ErrorDisplay error={error} />}
      <PrimaryButton
        testID="Login.CheckEmailBtn"
        title={t('login.form.btnTitle')}
        onPress={handleSubmit}
        loading={loading}
        // style={LoginStyles.formBtn}
      />
    </AnimatedView>
  );
};

export default Email;
