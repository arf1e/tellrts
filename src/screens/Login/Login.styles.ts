import StyleSheet from 'react-native-extended-stylesheet';

export default StyleSheet.create({
  background: {
    paddingHorizontal: 24,
    paddingVertical: 24,
    paddingTop: '20%',
    backgroundColor: '$primary',
    flex: 1,
  },

  logo: {
    width: 300,
    marginBottom: 40,
    alignSelf: 'center',
  },

  title: {
    color: '$background',
    fontSize: '$h1',
  },

  formContainer: {
    backgroundColor: '$background',
    padding: 16,
    paddingTop: 20,
    borderRadius: 5,
    shadowColor: '$darkGray',
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.2,
    shadowRadius: 1.41,

    elevation: 2,
  },

  formTitle: {
    fontSize: '$h4',
    marginBottom: 32,
  },

  fieldTitle: {
    marginBottom: 7,
  },

  emailField: {
    marginBottom: 16,
  },

  emailBtn: {
    color: '$primary',
    fontSize: '$bcMedium',
    fontFamily: 'Roboto-Regular',
    marginBottom: 16,
  },
});
