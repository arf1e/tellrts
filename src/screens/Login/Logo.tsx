import {Image} from 'react-native';
