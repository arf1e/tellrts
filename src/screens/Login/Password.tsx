import {useNavigation} from '@react-navigation/native';
import React from 'react';
import {useTranslation} from 'react-i18next';
import {View} from 'react-native';
import {Formik} from 'formik';
import * as yup from 'yup';

import PrimaryButton from '../../components/Buttons';
import {SIGNUP_SCREEN} from '../../components/Navigation/AuthNavigator';
import {BodyCopy} from '../../components/Typography';
import Field from '../../components/Field';
import LoginStyles from './Login.styles';
import {gql, useMutation} from '@apollo/client';
import {useDispatch} from 'react-redux';
import {logIn} from '../../utils/slices/authSlice';
import errorCatcher from '../../utils/toasts';

interface Props {
  emailExists: boolean;
  email: string;
  goBack: () => void;
}

const passwordValidationSchema = yup.object({
  password: yup.string().required(),
});

const initialValues = {
  password: '',
};

const LOGIN_MUTATION = gql`
  mutation login($email: String!, $password: String!) {
    login(email: $email, password: $password) {
      ok
      error
      token
    }
  }
`;

const Password = ({emailExists, email, goBack}: Props) => {
  const {t} = useTranslation();
  const navigation = useNavigation();

  const dispatch = useDispatch();

  const onLoginMutationCompleted = ({
    error,
    token,
  }: {
    error: string;
    token: string;
  }) => {
    if (error) {
      errorCatcher(error);
      return;
    }

    if (token) {
      dispatch(logIn({token}));
      return;
    }
  };

  const [login, {data, loading, error}] = useMutation(LOGIN_MUTATION, {
    onCompleted: loginMutationData =>
      onLoginMutationCompleted(loginMutationData.login),
    onError: errorCatcher,
  });

  const handleLogin = (password: string) => {
    login({variables: {email, password}});
  };

  const renderSignUpButton = () => (
    <PrimaryButton
      title={t('login.form.signUpBtn')}
      onPress={() => navigation.navigate(SIGNUP_SCREEN, {email})}
    />
  );

  const renderLoginForm = () => {
    return (
      <>
        <BodyCopy style={LoginStyles.fieldTitle}>
          {t('login.form.passwordFieldTitle')}
        </BodyCopy>
        <Formik
          initialValues={initialValues}
          validationSchema={passwordValidationSchema}
          onSubmit={values => handleLogin(values.password)}>
          {({handleChange, handleSubmit}) => (
            <>
              <Field
                secureTextEntry
                autoFocus
                style={LoginStyles.emailField}
                onChangeText={handleChange('password')}
              />
              <PrimaryButton
                title={t('login.form.loginBtn')}
                onPress={handleSubmit}
              />
            </>
          )}
        </Formik>
      </>
    );
  };
  return (
    <View>
      <BodyCopy
        onPress={goBack}
        style={LoginStyles.emailBtn}>{`< ${email}`}</BodyCopy>
      {!emailExists && renderSignUpButton()}
      {emailExists && renderLoginForm()}
    </View>
  );
};

export default Password;
