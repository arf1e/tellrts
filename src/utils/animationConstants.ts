const animationConstants: {[key: string]: number} = {
  BUTTON_IN: 180,
  BUTTON_OUT: 160,
  SCALE_ON_PRESS: 0.98,
};

export default animationConstants;
