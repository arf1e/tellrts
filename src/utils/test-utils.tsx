import React, {ReactNode} from 'react';
import {MockedProvider} from '@apollo/client/testing';
import {Provider} from 'react-redux';
import {NavigationContainer} from '@react-navigation/native';
import {configureStore} from '@reduxjs/toolkit';
import authSliceReducer from './slices/authSlice';
import firebaseTokenSliceReducer from './slices/firebaseTokenSlice';
import anketSliceReducer from './slices/anketSlice';
import requestResultSliceReducer from './slices/requestResultSlice';
import inChatSliceReducer from './slices/inChatSlice';
import contactsInputSliceReducer from './slices/contactsInputSlice';
import searchSettingsSliceReducer from './slices/searchSettingsSlice';

const reducer = {
  auth: authSliceReducer,
  firebase: firebaseTokenSliceReducer,
  anket: anketSliceReducer,
  requestResult: requestResultSliceReducer,
  inChat: inChatSliceReducer,
  contactsInput: contactsInputSliceReducer,
  searchSettings: searchSettingsSliceReducer,
};

export function renderWithProviders(
  ui: ReactNode,
  {
    preloadedState = {},
    mocks = [],
    store = configureStore({reducer, preloadedState}),
    ...renderOptions
  } = {},
) {
  return (
    <MockedProvider mocks={mocks}>
      <Provider store={store}>
        <NavigationContainer>{ui} </NavigationContainer>
      </Provider>
    </MockedProvider>
  );
}
