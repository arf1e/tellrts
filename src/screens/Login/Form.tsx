import React, {useState} from 'react';
import {useTranslation} from 'react-i18next';
import {View} from 'react-native';
import {Formik} from 'formik';
import * as yup from 'yup';
import {gql, useLazyQuery} from '@apollo/client';
import Reanimated, {FadeInUp, Layout} from 'react-native-reanimated';

import {Subtitle} from '../../components/Typography';
import LoginStyles from './Login.styles';
import Email from './Email';
import Password from './Password';
import errorCatcher from '../../utils/toasts';

const AnimatedView = Reanimated.createAnimatedComponent(View);

export const CHECK_EMAIL_QUERY = gql`
  query emailExists($email: String!) {
    checkEmail(email: $email) {
      ok
      emailExists
    }
  }
`;

const schema = yup.object({
  email: yup.string().required().email('login.email.errors.incorrectEmail'),
});

const initialValues = {
  email: '',
};

const Form = () => {
  const {t} = useTranslation();
  const [step, setStep] = useState<0 | 1>(0);
  const [checkEmail, {data, loading: checkEmailLoading}] = useLazyQuery(
    CHECK_EMAIL_QUERY,
    {
      fetchPolicy: 'network-only',
      onCompleted: () => {
        setStep(1);
      },
      onError: errorCatcher,
    },
  );
  return (
    <AnimatedView
      entering={FadeInUp.duration(240).damping(240)}
      layout={Layout}
      style={LoginStyles.formContainer}>
      <Subtitle style={LoginStyles.formTitle}>{t('login.form.title')}</Subtitle>
      <Formik
        initialValues={initialValues}
        onSubmit={({email}) => {
          console.log({email});
          // await checkEmail({variables: {email}});
        }}
        validationSchema={schema}
        validateOnBlur={true}
        validateOnChange={false}>
        {({handleChange, handleSubmit, errors, values}) => (
          <>
            {step === 0 && (
              <Email
                value={values.email}
                isInvalid={Boolean(errors.email)}
                error={errors.email}
                handleChange={handleChange('email')}
                handleSubmit={handleSubmit}
                loading={checkEmailLoading}
              />
            )}
            {step === 1 && (
              <Password
                email={values.email}
                emailExists={data?.checkEmail?.emailExists}
                goBack={() => setStep(0)}
              />
            )}
          </>
        )}
      </Formik>
    </AnimatedView>
  );
};

export default Form;
