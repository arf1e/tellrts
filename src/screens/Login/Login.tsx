import React from 'react';
import {Image, StatusBar, View} from 'react-native';
import LinearGradient from 'react-native-linear-gradient';
import DismissKeyboard from '../../components/DismissKeyboard';
import colors from '../../utils/colors';
import Form from './Form';
import styles from './Login.styles';

const Login = () => {
  return (
    <DismissKeyboard>
      <LinearGradient
        colors={['#1693A5', '#6C93BE', '#DC8DCB']}
        style={styles.background}
        start={{x: 0, y: 0}}
        end={{x: 1, y: 1}}>
        <View>
          <StatusBar backgroundColor={colors.primary} />
          <Image
            source={require('../../../assets/img/logo.png')}
            style={styles.logo}
          />
          <Form />
        </View>
      </LinearGradient>
    </DismissKeyboard>
  );
};

export default Login;
