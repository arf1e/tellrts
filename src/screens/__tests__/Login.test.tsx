import {render} from '@testing-library/react-native';
import React from 'react';
import Login from '../Login';
import {renderWithProviders} from '../../utils/test-utils';
beforeAll(() => {
  jest.mock('react-native-reanimated', () =>
    jest.requireActual(
      '../../node_modules/react-native-reanimated/lib/reanimated2/mock',
    ),
  );
});

test('Login Screen Renders', () => {
  const view = render(renderWithProviders(<Login />, {}));
  expect(view).toMatchSnapshot();
});
