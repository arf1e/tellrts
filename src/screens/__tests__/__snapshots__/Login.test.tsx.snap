// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Screen Renders 1`] = `
Array [
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <View>
      <Image
        resizeMode="contain"
        source={
          Object {
            "testUri": "../../../assets/img/logo.png",
          }
        }
      />
      <View
        animatedStyle={
          Object {
            "value": Object {},
          }
        }
        collapsable={false}
        entering={
          FadeInUp {
            "build": [Function],
            "dampingV": 240,
            "durationV": 240,
            "randomizeDelay": false,
          }
        }
        layout={[Function]}
        style={Object {}}
      >
        <Text
          style={
            Array [
              undefined,
              undefined,
            ]
          }
        >
          login.form.title
        </Text>
        <View
          animatedStyle={
            Object {
              "value": Object {},
            }
          }
          collapsable={false}
          entering={
            FadeInLeft {
              "build": [Function],
              "delayV": 120,
              "durationV": 320,
              "randomizeDelay": false,
            }
          }
          exiting={
            FadeOutLeft {
              "build": [Function],
              "dampingV": 5000,
              "durationV": 320,
              "randomizeDelay": false,
            }
          }
        >
          <Text
            style={
              Array [
                undefined,
                undefined,
              ]
            }
          >
            login.form.emailFieldTitle
          </Text>
          <TextInput
            animatedStyle={
              Object {
                "value": Object {
                  "borderColor": "rgba(188, 216, 221, 1)",
                },
              }
            }
            autoCapitalize="none"
            collapsable={false}
            defaultValue=""
            keyboardType="email-address"
            onBlur={[Function]}
            onChangeText={[Function]}
            onFocus={[Function]}
            placeholder="login.form.emailPlaceholder"
            style={
              Object {
                "borderColor": "rgba(188, 216, 221, 1)",
              }
            }
            testID="Field.TextInput"
          />
          <View
            accessibilityState={
              Object {
                "disabled": false,
              }
            }
            accessible={true}
            animatedStyle={
              Object {
                "value": Object {
                  "backgroundColor": "rgba(22, 147, 165, 1)",
                  "transform": Array [
                    Object {
                      "scale": 1,
                    },
                  ],
                },
              }
            }
            collapsable={false}
            focusable={true}
            loading={false}
            onBlur={[Function]}
            onClick={[Function]}
            onFocus={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "backgroundColor": "rgba(22, 147, 165, 1)",
                "transform": Array [
                  Object {
                    "scale": 1,
                  },
                ],
              }
            }
            testID="PrimaryButton.container"
          >
            <Text
              animatedStyle={
                Object {
                  "value": Object {
                    "opacity": 1,
                    "transform": Array [
                      Object {
                        "translateY": 0,
                      },
                    ],
                  },
                }
              }
              collapsable={false}
              style={
                Object {
                  "opacity": 1,
                  "transform": Array [
                    Object {
                      "translateY": 0,
                    },
                  ],
                }
              }
            >
              login.form.btnTitle
            </Text>
            <View
              animatedStyle={
                Object {
                  "value": Object {
                    "display": "none",
                    "opacity": 0,
                    "position": "absolute",
                    "transform": Array [
                      Object {
                        "translateY": 20,
                      },
                    ],
                  },
                }
              }
              collapsable={false}
              style={
                Object {
                  "display": "none",
                  "opacity": 0,
                  "position": "absolute",
                  "transform": Array [
                    Object {
                      "translateY": 20,
                    },
                  ],
                }
              }
            >
              <ActivityIndicator
                color="#FDFFFF"
                size={24}
              />
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>,
  " ",
]
`;
